File: src/com/company/boj/dfs/BOJ_6603_로또.java
package com.company.boj.dfs;

import java.util.Scanner;

/**
 * Created by JihoKevin.
 * User: sinjiho
 * Date: 2022/02/21
 * Time: 3:54 오후
 */
public class BOJ_6603_로또 {

    public static void main(String[] args) {

        Scanner scan = new Scanner(System.in);

        int[] lotto;
        boolean[] visited;
        while (true) {
            int k = scan.nextInt();

            if (k == 0) {
                break;
            }
            lotto = new int[k];
            visited = new boolean[k];
            for (int i = 0; i < k; i++) {
                lotto[i] = scan.nextInt();
            }
            System.out.println();
            dfs(0, 0, lotto, visited);
        }
    }

    public static void dfs(int start, int depth, int[] lotto, boolean[] visited) {

        if (depth == 6) {
            for (int i = 0; i < lotto.length; i++) {
                if (visited[i]) System.out.print(lotto[i] + " ");
            }
            System.out.println();
        }

        for (int i = start; i < lotto.length; i++) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(i + 1, depth + 1, lotto, visited);
                visited[i] = false;
            }
        }
    }
}

File: src/com/company/boj/dfs/BOJ_6603_로또.kt
package com.company.boj.dfs

import java.util.*

/**
 * Created by JihoKevin.
 * User: sinjiho
 * Date: 2022/02/26
 * Time: 12:56 오전
 */

private var n = 0
private var lotto = intArrayOf()
private var visited = booleanArrayOf()

fun main(args: Array<String>) {

    val sc: Scanner = Scanner(System.`in`)

    while (true) {
        n = sc.nextInt()
        if (n == 0) break
        lotto = IntArray(n)
        visited = BooleanArray(n)
        for (i in 0 until n) {
            lotto[i] = sc.nextInt()
        }
        println()
        dfs(0, 0, lotto, visited)
    }
}

fun dfs(start: Int, depth: Int, lotto: IntArray, visited: BooleanArray) {

    if (depth == 6) {
        for (i in lotto.indices) {
            if (visited[i]) print(lotto[i])
        }
        println()
    }

    for (i in start until n) {
        if (!visited[i]) {
            visited[i] = true;
            dfs(i + 1, depth + 1, lotto, visited)
            visited[i] = false;
        }
    }
}